symengine/tests/basic/test_solve.cpp

```diff
@@ -0,0 +1,372 @@
+using SymEngine::NegInf;
+using SymEngine::I;
+using SymEngine::SymEngineException;
+using SymEngine::neg;
+using SymEngine::one;
+using SymEngine::zero;
+using SymEngine::pow;
+using SymEngine::FiniteSet;
+using SymEngine::Eq;
+using SymEngine::Ne;
+using SymEngine::Ge;
+using SymEngine::boolTrue;
+using SymEngine::boolFalse;
+using SymEngine::down_cast;
+using SymEngine::ConditionSet;
+using SymEngine::is_a;
+using SymEngine::logical_and;
+using SymEngine::Union;
+using SymEngine::mul;
+using SymEngine::UIntPoly;
+using SymEngine::URatPoly;
+using SymEngine::rational_class;
+#ifdef HAVE_SYMENGINE_FLINT
+using SymEngine::UIntPolyFlint;
+using SymEngine::URatPolyFlint;
+#endif
+#ifdef HAVE_SYMENGINE_PIRANHA
+using SymEngine::UIntPolyPiranha;
+using SymEngine::URatPolyPiranha;
+#endif
+using namespace SymEngine::literals;
+
+TEST_CASE("Trivial cases", "[Solve]")
+{
+    RCP<const Symbol> x = symbol("x");
+    RCP<const Basic> poly;
+    RCP<const Set> reals = interval(NegInf, Inf, true, true);
+    RCP<const Set> soln;
+
+    REQUIRE(eq(*solve(boolTrue, x, reals), *reals));
+    REQUIRE(eq(*solve(boolFalse, x, reals), *emptyset()));
+
+    // constants
+    poly = one;
+    soln = solve(poly, x, reals);
+    REQUIRE(eq(*soln, *emptyset()));
+
+    poly = zero;
+    soln = solve(poly, x, reals);
+    REQUIRE(eq(*soln, *reals));
+}
+
+TEST_CASE("linear and quadratic polynomials", "[Solve]")
+{
+    RCP<const Symbol> x = symbol("x");
+    RCP<const Basic> poly;
+    RCP<const Set> reals = interval(NegInf, Inf, true, true);
+    RCP<const Set> soln;
+
+    auto i2 = integer(2), i3 = integer(3), im2 = integer(-2), im3 = integer(-3);
+    auto sqx = mul(x, x);
+
+    // linear
+    poly = add(x, i3);
+    soln = solve(poly, x, reals);
+    REQUIRE(eq(*soln, *finiteset({im3})));
+
+    soln = solve(poly, x, interval(zero, Inf, false, true));
+    REQUIRE(eq(*soln, *emptyset()));
+
+    poly = add(div(x, i2), div(one, i3));
+    soln = solve(poly, x);
+    REQUIRE(eq(*soln, *finiteset({neg(rational(2, 3))})));
+
+    poly = x;
+    soln = solve(poly, x, reals);
+    REQUIRE(eq(*soln, *finiteset({zero})));
+
+    auto y = symbol("y");
+    soln = solve(add(x, y), x);
+    REQUIRE(eq(*soln, *finiteset({neg(y)})));
+
+    CHECK_THROWS_AS(solve_poly_linear({one}, x, reals), SymEngineException);
+
+    // Quadratic
+    poly = add(sqx, one);
+    soln = solve(poly, x);
+    REQUIRE(eq(*soln, *finiteset({neg(I), I})));
+
+    poly = add(add(sqx, mul(im2, x)), one);
+    soln = solve(poly, x);
+    REQUIRE(eq(*soln, *finiteset({one})));
+
+    poly = sub(add(div(sqx, i3), x), div(i2, integer(5)));
+    soln = solve(poly, x);
+    REQUIRE(
+        eq(*soln,
+           *finiteset(
+               {add(div(mul(sqrt(integer(69)), sqrt(integer(5))), integer(10)),
+                    div(im3, i2)),
+                sub(div(im3, i2), div(mul(sqrt(integer(69)), sqrt(integer(5))),
+                                      integer(10)))}))); // Shouldn't
```

Comment:
Why is `sqrt(5)` not a number (It should be)?  Looking at `pow` in `pow.cpp` there is a case where both 5 and 1/2 are numbers, and the result's output should be a number. Can you check?